src/parser/if-then-else.js

```diff
@@ -4,47 +4,53 @@ import type Context from "./context";
 import statement from "./statement";
 import expression from "./expression";
 
-const ifThenElse = (ctx: Context) => {
-  const node = {
-    ...ctx.startNode(ctx.token),
-    then: [],
-    else: []
-  };
-  ctx.eat(["if"]);
-  // First operand is the expression
+// push statement based on then/else branch
+const pushStatement = (node, isThenBranch, stmt) => {
+  if (!stmt) return;
+  if (isThenBranch) return node.then.push(stmt);
+  return node.else.push(stmt);
+}
+
+const doIfExpression = (node, ctx) => {
   ctx.expect(["("]);
   node.expr = expression(ctx, "i32");
   ctx.expect([")"]);
+}
 
-  // maybe a curly brace or not
+// push statements while taking into consideration having curly braces or not
+const doStatement = (node, isThenBranch, ctx) => {
   if (ctx.eat(["{"])) {
-    let stmt = null;
     while (ctx.token && ctx.token.value !== "}") {
-      stmt = statement(ctx);
-      if (stmt) {
-node.then.push(stmt);
-}
+      pushStatement(node, isThenBranch, statement(ctx));
     }
-
     ctx.expect(["}"]);
-
-    if (ctx.eat(["else"])) {
-      ctx.expect(["{"]);
-      while (ctx.token && ctx.token.value !== "}") {
-        stmt = statement(ctx);
-        if (stmt) {
-node.else.push(stmt);
-}
-      }
-      ctx.expect(["}"]);
-    }
   } else {
-    // parse single statements only
-    node.then.push(statement(ctx));
+    pushStatement(node, isThenBranch, statement(ctx));
     ctx.expect([";"]);
-    if (ctx.eat(["else"])) {
-node.else.push(statement(ctx));
+  }
 }
+
+const ifThenElse = (ctx: Context) => {
+  const node = {
+    ...ctx.startNode(ctx.token),
+    then: [],
+    else: []
+  };
+  let isThenBranch = true;
+
+  ctx.eat(["if"]);
+  // First operand is the expression
+  doIfExpression(node, ctx);
+  doStatement(node, isThenBranch, ctx);
+
+  while (ctx.eat(["else"])) {
+    isThenBranch = false;
+    // maybe another if statement
+    if (ctx.eat(["if"])) {
+      isThenBranch = true;
+      doIfExpression(node, ctx);
+    }
+    doStatement(node, isThenBranch, ctx);
```

Comment:
I don't think we need `isThenBranch` or `pushStatement`. Let's just simplify a bit by

```js
isThenBranch = ctx.eat(['if']);
if (ctx.eat(['if'])) {
   doIfExpression(node, ctx);
   node.then = [...node.then, doStatement(ctx)];
} else {
   node.else = [ ...node.else, doStatement(ctx)];
}
```

Reply:
Agreed, i had an uglier implementation before that worked. Your ideea is better. I'll update the PR with the changes (test cases for all branches as well)